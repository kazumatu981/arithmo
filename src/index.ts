export function sayHello(_: string): void {}
