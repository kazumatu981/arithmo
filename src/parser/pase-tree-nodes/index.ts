export * from './parse-tree-node';
export * from './single-node';
export * from './binary-node';
export * from './paren-node';
