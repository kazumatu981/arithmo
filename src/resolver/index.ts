export * from './number-resolver';
export * from './common/resolver-base';
