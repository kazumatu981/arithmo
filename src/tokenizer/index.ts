import { type Token, type TokenType } from './token';
import { Tokenizer } from './tokenizer';

/**
 * 入力文字列を字句に分割する
 * @param input 入力文字列
 * @returns 切り出した字句
 */
function tokenize(input: string): Token[] {
    return new Tokenizer(input).tokenize();
}

export { type Token, type TokenType, tokenize };
