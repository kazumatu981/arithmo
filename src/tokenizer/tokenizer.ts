import { TokenizerError } from '../common/errors';
import {
    isDigit,
    isOperator,
    isParen,
    isWhiteSpace,
    isLeftParen,
} from '../common/char-util';
import { Token } from './token';

export class Tokenizer {
    private expression: string;
    private currentIndex = 0;
    constructor(expression: string) {
        this.expression = expression;
    }

    /**
     * 入力文字列を字句に分割する
     * @returns 切り出した字句
     */
    public tokenize(): Token[] {
        this.reset();
        const tokens: Token[] = [];
        const tokenIterator = this.tokens();
        for (const token of tokenIterator) {
            tokens.push(token);
        }
        return tokens;
    }

    public *tokens(): Generator<Token> {
        this.reset();
        while (this.currentIndex < this.expression.length) {
            const token = this.readNextToken();
            if (token) {
                yield token;
            }
        }
    }
    public reset(): void {
        this.currentIndex = 0;
    }

    /**
     * inputのstartIndexから始まる字句を切り出す。
     * @param input 入力文字列
     * @param startIndex 開始位置
     * @returns 切り切り出した字句と次の開始位置
     */
    public readNextToken(): Token | undefined {
        this._skipSpaces();

        if (this.currentIndex >= this.expression.length) {
            // 空白文字のまま末尾まで行ったらトークンは何もなかったとみなす。
            return undefined;
        } else if (isDigit(this.expression, this.currentIndex)) {
            // 数字を切り出す
            return this._readNumberToken();
        } else if (isOperator(this.expression, this.currentIndex)) {
            // 演算子を切り出す
            return this._readOperatorToken();
        } else if (isParen(this.expression, this.currentIndex)) {
            // 括弧を切り出す
            return this._readParenToken();
        } else {
            // 予期せぬ文字を検出した
            throw new TokenizerError(
                'unknown-character',
                '',
                this.currentIndex,
            );
        }
    }

    /**
     * 空白文字をスキップする
     */
    private _skipSpaces(): void {
        while (
            this.currentIndex < this.expression.length &&
            isWhiteSpace(this.expression, this.currentIndex)
        ) {
            this.currentIndex++;
        }
    }

    /**
     * 数字文字列としてTokenを切り出す。
     * @returns 切り出した字句
     */
    private _readNumberToken(): Token {
        let value = '';
        const startIndex = this.currentIndex;
        while (
            this.currentIndex < this.expression.length &&
            isDigit(this.expression, this.currentIndex)
        ) {
            value += this.expression.charAt(this.currentIndex);
            this.currentIndex++;
        }
        return new Token('number', value, startIndex);
    }

    /**
     * 演算子文字列としてTokenを切り出す。
     * @returns 切り出した字句
     */
    private _readOperatorToken(): Token {
        const startIndex = this.currentIndex;
        this.currentIndex++;
        return new Token(
            'operator',
            this.expression.charAt(startIndex),
            startIndex,
        );
    }

    /**
     * 括弧文字列としてTokenを切り出す。
     * @returns 切り出した字句
     */
    private _readParenToken(): Token {
        const startIndex = this.currentIndex;
        const parenCharacter = this.expression.charAt(startIndex);
        this.currentIndex++;
        return new Token(
            isLeftParen(this.expression, startIndex)
                ? 'leftParen'
                : 'rightParen',
            parenCharacter,
            startIndex,
        );
    }
}
