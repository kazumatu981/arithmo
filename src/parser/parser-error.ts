import {
    ArithmoError,
    type ErrorCode,
    type ErrorOptions,
} from '../common/errors';

import { type Token } from '../tokenizer';

/**
 * 構文解析エラーのオプション
 */
export interface ParserErrorOptions extends ErrorOptions {
    /**
     * 発生した字句
     */
    token?: Token;
}

/**
 * 構文解析エラー
 */
export class ParserError extends ArithmoError {
    public readonly token?: Token;

    /**
     * 構文解析時に発生したエラー
     * @param code - エラーコード
     * @param options - エラーオプション
     */
    public constructor(code: ErrorCode, options?: ParserErrorOptions) {
        if (options && options.position === undefined) {
            options.position = options?.token?.position;
        }
        super(code, 'parser', options);
        this.token = options?.token;
    }
}
