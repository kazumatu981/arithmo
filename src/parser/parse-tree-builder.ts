import type { Token, TokenType } from '../tokenizer';
import {
    ParseTreeNode,
    SingleNode,
    BinaryNode,
    ParenNode,
} from './pase-tree-nodes';
import type { ErrorCode } from '../common/error-messages';
import { ParserError } from './parser-error';

type ParseState = 'Initial' | 'WaitForNumber' | 'WaitForOperator';

type ParseTreeStateTableRecord<T> = Record<TokenType, T>;
type ParseTreeStateTable<T> = Record<ParseState, ParseTreeStateTableRecord<T>>;

type StateAction = (token: Token) => void;

export class ParseTreeBuilder {
    private _currentNode?: ParseTreeNode;
    private state: ParseState = 'Initial';
    private sign: Token | null = null;

    private nextStateTable: ParseTreeStateTable<ParseState | undefined> = {
        Initial: {
            number: 'WaitForOperator',
            leftParen: 'Initial',
            rightParen: undefined,
            operator: 'WaitForNumber',
        },
        WaitForNumber: {
            number: 'WaitForOperator',
            leftParen: 'Initial',
            rightParen: undefined,
            operator: undefined,
        },
        WaitForOperator: {
            number: undefined,
            leftParen: undefined,
            rightParen: 'WaitForOperator',
            operator: 'WaitForNumber',
        },
    };

    private stateActionTable: ParseTreeStateTable<StateAction | undefined> = {
        Initial: {
            number: this.appendNumberNode.bind(this),
            leftParen: this.appendParenStart.bind(this),
            rightParen: ((token: Token): void => {
                this.throwParseError('unexpected-right-paren', token);
            }).bind(this),
            operator: this.appendSign.bind(this),
        },
        WaitForNumber: {
            number: this.appendNumberNode.bind(this),
            leftParen: this.appendParenStart.bind(this),
            rightParen: ((token: Token): void => {
                this.throwParseError('unexpected-right-paren', token);
            }).bind(this),
            operator: ((token: Token): void => {
                this.throwParseError('unexpected-operator', token);
            }).bind(this),
        },
        WaitForOperator: {
            number: ((token: Token): void => {
                this.throwParseError('unexpected-number', token);
            }).bind(this),
            leftParen: ((token: Token): void => {
                this.throwParseError('unexpected-left-paren', token);
            }).bind(this),
            rightParen: this.appendParenEnd.bind(this),
            operator: this.appendOperatorNode.bind(this),
        },
    };
    public addToken(token: Token): this {
        const action = this.stateActionTable[this.state][token.type];
        const nextState = this.nextStateTable[this.state][token.type];
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        action!(token);
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        this.state = nextState!;
        return this;
    }
    public findRootNode(): ParseTreeNode | undefined {
        return ParseTreeNode.findRootNode(this._currentNode);
    }
    public build(): ParseTreeNode | undefined {
        return this.findRootNode()?.validate();
    }

    private throwParseError(code: ErrorCode, token: Token): void {
        throw new ParserError(code, { token });
    }
    private appendSign(token: Token): void {
        if (token.isNegativeSign) {
            this.sign = token;
        } else {
            throw new ParserError('unexpected-sign', { token });
        }
    }

    private appendNumberNode(token: Token): void {
        const numberNode = this.sign
            ? new SingleNode([this.sign, token])
            : new SingleNode([token]);
        this.sign = null;
        if (this._currentNode) {
            if (this._currentNode.nodeType === 'binary') {
                (this._currentNode as BinaryNode).right = numberNode;
            } else if (this._currentNode.nodeType === 'paren') {
                (this._currentNode as ParenNode).childrenRoot = numberNode;
            } else {
                throw new ParserError('unexpected', {
                    token,
                    appendixMessage:
                        '現在のノードに数字ノードを追加しようとたが、子ノードを持てないノードである(バグの可能性がある)',
                });
            }
        }
        this._currentNode = numberNode;
    }
    private appendOperatorNode(token: Token): void {
        this._currentNode = new BinaryNode([token]).attachTo(this._currentNode);
    }
    private appendParenStart(token: Token): void {
        const parenNode = this.sign
            ? new ParenNode([this.sign, token])
            : new ParenNode([token]);
        this.sign = null;
        if (this._currentNode) {
            if (this._currentNode.nodeType === 'binary') {
                (this._currentNode as BinaryNode).right = parenNode;
            } else if (this._currentNode.nodeType === 'paren') {
                (this._currentNode as ParenNode).childrenRoot = parenNode;
            } else {
                throw new ParserError('unexpected', {
                    token,
                    appendixMessage:
                        '現在のノードに括弧ノードを追加しようとたが、子ノードを持てないノードである(バグの可能性がある)',
                });
            }
        }
        this._currentNode = parenNode;
    }
    private appendParenEnd(token: Token): void {
        const parenNode = ParenNode.findParenNode(this._currentNode);
        if (!parenNode) {
            throw new ParserError('unexpected-right-paren', { token });
        }
        parenNode.parenEnd = token;
        this._currentNode = parenNode;
    }
}
