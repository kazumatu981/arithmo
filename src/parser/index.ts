import { type Token } from '../tokenizer';
import { ParseTreeBuilder } from './parse-tree-builder';
import { type ParseTreeNode } from './parse-tree-nodes';

export * from './parse-tree-nodes';
export function parse(tokens: Token[]): ParseTreeNode | undefined {
    const builder = new ParseTreeBuilder();
    tokens.forEach((token) => builder.addToken(token));
    return builder.build();
}
