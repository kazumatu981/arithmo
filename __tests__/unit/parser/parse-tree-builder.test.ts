import { describe, expect, test } from '@jest/globals';
import { ParseTreeBuilder } from '../../../src/parser/parse-tree-builder';
import { Tokenizer } from '../../../src/tokenizer';

const testCases: string[] = ['1+1', '1-1', '1*12+1', '1*12+3*4'];
describe('unittest-parse-tree-builder', () => {
    testCases.forEach((testCase) => {
        test(`${testCase}`, () => {
            const tokenizer = new Tokenizer(testCase);
            const tokens = tokenizer.tokenize();
            const builder = new ParseTreeBuilder();
            tokens.forEach((token) => builder.addToken(token));
            const tree = builder.build();
            expect(tree?.toNodeInfo()).toMatchSnapshot();
        });
    });
});
