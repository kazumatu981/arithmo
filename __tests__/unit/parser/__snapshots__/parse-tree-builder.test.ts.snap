// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unittest-parse-tree-builder Can build tree from (1+1)*2 1`] = `
{
  "left": {
    "childrenRoot": {
      "left": {
        "type": "single",
        "value": "1",
      },
      "right": {
        "type": "single",
        "value": "1",
      },
      "type": "binary",
      "value": "+",
    },
    "type": "paren",
    "value": "()",
  },
  "right": {
    "type": "single",
    "value": "2",
  },
  "type": "binary",
  "value": "*",
}
`;

exports[`unittest-parse-tree-builder Can build tree from -(1-(-1+5)) 1`] = `
{
  "childrenRoot": {
    "left": {
      "type": "single",
      "value": "1",
    },
    "right": {
      "childrenRoot": {
        "left": {
          "type": "single",
          "value": "-1",
        },
        "right": {
          "type": "single",
          "value": "5",
        },
        "type": "binary",
        "value": "+",
      },
      "type": "paren",
      "value": "()",
    },
    "type": "binary",
    "value": "-",
  },
  "type": "paren",
  "value": "-()",
}
`;

exports[`unittest-parse-tree-builder Can build tree from -1+1 1`] = `
{
  "left": {
    "type": "single",
    "value": "-1",
  },
  "right": {
    "type": "single",
    "value": "1",
  },
  "type": "binary",
  "value": "+",
}
`;

exports[`unittest-parse-tree-builder Can build tree from 1*12+1 1`] = `
{
  "left": {
    "left": {
      "type": "single",
      "value": "1",
    },
    "right": {
      "type": "single",
      "value": "12",
    },
    "type": "binary",
    "value": "*",
  },
  "right": {
    "type": "single",
    "value": "1",
  },
  "type": "binary",
  "value": "+",
}
`;

exports[`unittest-parse-tree-builder Can build tree from 1*12+3*4 1`] = `
{
  "left": {
    "left": {
      "type": "single",
      "value": "1",
    },
    "right": {
      "type": "single",
      "value": "12",
    },
    "type": "binary",
    "value": "*",
  },
  "right": {
    "left": {
      "type": "single",
      "value": "3",
    },
    "right": {
      "type": "single",
      "value": "4",
    },
    "type": "binary",
    "value": "*",
  },
  "type": "binary",
  "value": "+",
}
`;

exports[`unittest-parse-tree-builder Can build tree from 1+(-1) 1`] = `
{
  "left": {
    "type": "single",
    "value": "1",
  },
  "right": {
    "childrenRoot": {
      "type": "single",
      "value": "-1",
    },
    "type": "paren",
    "value": "()",
  },
  "type": "binary",
  "value": "+",
}
`;

exports[`unittest-parse-tree-builder Can build tree from 1+(3*3)+1 1`] = `
{
  "left": {
    "left": {
      "type": "single",
      "value": "1",
    },
    "right": {
      "childrenRoot": {
        "left": {
          "type": "single",
          "value": "3",
        },
        "right": {
          "type": "single",
          "value": "3",
        },
        "type": "binary",
        "value": "*",
      },
      "type": "paren",
      "value": "()",
    },
    "type": "binary",
    "value": "+",
  },
  "right": {
    "type": "single",
    "value": "1",
  },
  "type": "binary",
  "value": "+",
}
`;

exports[`unittest-parse-tree-builder Can build tree from 1+1 1`] = `
{
  "left": {
    "type": "single",
    "value": "1",
  },
  "right": {
    "type": "single",
    "value": "1",
  },
  "type": "binary",
  "value": "+",
}
`;

exports[`unittest-parse-tree-builder Can build tree from 1-(-1+5) 1`] = `
{
  "left": {
    "type": "single",
    "value": "1",
  },
  "right": {
    "childrenRoot": {
      "left": {
        "type": "single",
        "value": "-1",
      },
      "right": {
        "type": "single",
        "value": "5",
      },
      "type": "binary",
      "value": "+",
    },
    "type": "paren",
    "value": "()",
  },
  "type": "binary",
  "value": "-",
}
`;

exports[`unittest-parse-tree-builder Can build tree from 1-1 1`] = `
{
  "left": {
    "type": "single",
    "value": "1",
  },
  "right": {
    "type": "single",
    "value": "1",
  },
  "type": "binary",
  "value": "-",
}
`;

exports[`unittest-parse-tree-builder Cannot build tree from -(1 1`] = `"括弧ノードは閉じられる必要があります。"`;

exports[`unittest-parse-tree-builder Cannot build tree from 1 1 1`] = `"予期せぬ数字を検出しました。"`;

exports[`unittest-parse-tree-builder Cannot build tree from 1(1+3) 1`] = `"予期せぬ左括弧を検出しました。"`;

exports[`unittest-parse-tree-builder Cannot build tree from 1*(3*4 1`] = `"括弧ノードは2または3の字句を持つ必要があります。"`;

exports[`unittest-parse-tree-builder Cannot build tree from 1+ 1`] = `"演算子ノードに右の子ノードがありません。"`;

exports[`unittest-parse-tree-builder Cannot build tree from 1+() 1`] = `"予期せぬ右括弧を検出しました。"`;

exports[`unittest-parse-tree-builder Cannot build tree from 1+(3)) 1`] = `"予期せぬ右括弧を検出しました。"`;

exports[`unittest-parse-tree-builder Cannot build tree from 1--1-1 1`] = `"予期せぬ演算子を検出しました。"`;
